### PR TITLE
odc-grpc-server: add --infologger-severity

### LINKED ISSUE #92: odc-grpc-server: support configuring infoLogger severity separately

### PR DESCRIPTION
**Pull request by rbx**
Fixes #92.

### CODE REVIEW

File: odc/grpc/Controller.h [hunk 1]
```diff
@@ -455,9 +455,9 @@ class GrpcController final : public odc::ODC::Service
     {
         logGeneralReply(label, common, rep.reply());
         if (!rep.devices().empty()) {
-            OLOG(info, common) << "Detailed list of devices:";
+            OLOG(info, common) << "Detailed list of " << rep.devices().size() << " devices (logged with debug severity):";
```
Comment (davidrohr) [inline odc/grpc/Controller.h:458]: this should also be changed to debug, otherwise it doesn't make sense if the list does not follow

**Review by davidrohr — CHANGES_REQUESTED** [review]
And these messages should not be error during the startup
```


E | 2 | 10:27:04.428 | production | ODC | ODC | 2gXa57dSu1j |   | GetState  reply: partitionId: 2gXa57dSu1j; runnr: 0; sessionid:  7bf66742-f5cd-45ae-bb5d-c195c375f6a0; state: UNDEFINED; msg: ; exectime:  0ms; ERROR: Failed to get FairMQ device state (FairMQ topology is not  initialized) (302)
-- | -- | -- | -- | -- | -- | -- | -- | --
```
since it is normal behavior that some processes are undefined during startup.

**Review by davidrohr — COMMENTED** [review]
In addition, it would be good to fully make these messages `debug` in case they come from a GetState request:
```
I | 13 | 11:30:23.507 | epn-infra00 | production | ODC | ODC | 2gXcji4ZGXj |   | GetState  request for ODC 0.78.2 (DDS 3.7.22) from [ipv4:10.161.69.39:59638]  user-agent:grpc-go/1.50.1: partitionId: 2gXcji4ZGXj; runnr: 0; timeout: 0
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
I | 13 | 11:30:23.507 | epn-infra00 | production | ODC | ODC | 2gXcji4ZGXj |   | GetState request detailed: true; path:
I | 13 | 11:30:23.512 | epn-infra00 | production | ODC | ODC | 2gXcji4ZGXj |   | Device states: ERROR (2/22276) INITIALIZING DEVICE (6/22276) INITIALIZED (339/22276) RUNNING (21929/22276)
I | 13 | 11:30:23.512 | epn-infra00 | production | ODC | ODC | 2gXcji4ZGXj |   | Collection states: RUNNING (69/70) MIXED (1/70)
I | 13 | 11:30:23.516 | epn-infra00 | production | ODC | ODC | 2gXcji4ZGXj |   | GetState  reply: partitionId: 2gXcji4ZGXj; runnr: 0; sessionid:  4e741c5a-389a-4d90-9c57-d253e0d079ba; state: RUNNING; msg: GetState  done; exectime: 5ms
```
So for real state transitions / commands I'd like to have all this information, but perhaps not for the GetState request.

**Comment by rbx** [comment]
> since it is normal behavior that some processes are undefined during startup.

I agree. I would not set an error at all then.
If topology is not yet initialized I'll just return aggregatedState = Undefined, detailedState = []; No error.
Would that fit?

**Comment by rbx** [comment]
@lkrcal would changing GetState to debug logs be sufficient then? Or do you still want a --trace severity that would log the detailed device list as trace (which would not land in infoLogger at all, if --infologger-severity is debug)

**Comment by davidrohr** [comment]
> What about a case when Topology is in an Error state (I mean real error, nMin/expendables should be handled properly now)? GetState would return state = Error to ECS.
> 
> But currently GetState will not log and error in this case. Do you want it to log it as Error or just let controller handle the Error state. Such an error condition would already be logged by a different ODC component, e.g.:
> 
> ```
> 2023-07-19 10:26:38.670201 err  odc-grpc-server     1237096 1:0                  Collection 'main/SamplersSinks_0' (id: 2722891440665627871)has no nMin defined. Cannot be ignored.
> 2023-07-19 10:26:38.670244 err  odc-grpc-server     1237096 1:0                  Task 14311310895792912511 exited. Last known state: IDLE; path: "main/SamplersSinks/Sampler"; exit code: 137; signal: 9; host: witpc001; working directory: "/tmp/c6b6ad32-7694-4c86-87df-d277f8b1516a/dds_2023-07-19-10-20-28-555/wn/"
> ```
> 
> Not sure if logging also one more time with GetState at this point is useful. What do you think?

Your case makes sense. So if you can, I'd say GetState should log with Error severity, if the topology is not initialized, by the status is error.

**Comment by davidrohr** [comment]
> @lkrcal would changing GetState to debug logs be sufficient then? Or do you still want a --trace severity that would log the detailed device list as trace (which would not land in infoLogger at all, if --infologger-severity is debug)

Don't we meanwhile have a `detailed` severity for that, between info and debug, which does not go to infologger? At least we have it in O2, which also uses FairLogger.

**Comment by rbx** [comment]
ODC does not use FairLogger for historical reasons :).
But I can make another severity. If it's `detailed` or `trace`, is then a detail that we can also think about (could still map trace to something that infologger has, so it can be enabled if need be? not sure).
The question is, do you still need it now, or is the current case now solved by making GetState results debug only?

**Comment by davidrohr** [comment]
I agree that is a detail, and can be done later. For the GetState, I am not fully sure, but from my understanding, the getState function is also called in other state transitions, so it should not strictly be debug.
What I would do is:
- Printouts from an AliECS GetState command, while the topology is initialized, and not in error state, should be debug level in InfoLogger. In case of initialized topology and global error state, i.e. ODC in error, not individual tasks, I would indeed log with error severity.
- While the topology is not yet initialized, GetState should never log at error severity (this you solved by setting the state to undefined).
- Otherwise, it should be severity error in case of an error and info otherwise.

I went through the PR, and I think this is more or less what you implemented. Only I am not 100% sure about the calls to getState.

**Comment by rbx** [comment]
Yeah, the point from @lkrcal was that debug severity is still sent to infologger, even if you are not viewing it, so it will still take up space in the infologger DB (I can't judge how bad that is). With the current state of this PR you can now avoid that by setting --infologger-severity info in odc-grpc-server. If this is fine, then I'll merge it and tag.

Alternatively I can add trace severity, log the detailed device list in trace and then you will never see it in infologger, unless you set --infologger-sererity also to trace.

**Comment by davidrohr** [comment]
ok, but you replied yourself. We can and should just set the infologger-severity to info.

**Comment by rbx** [comment]
Alrighty, I'll merge it then and tag it later today.

**Comment by rbx** [comment]
This is now available with 0.79.0 tag.

**Pull request merged**